File: search a 2d Matrix 2.py
# TC-O(m+n)
# SC-O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n=len(matrix)
        m=len(matrix[0])
        # binary search
        # lowa=0
        # higha=m-1
        # for i in range(n):
        #     low=lowa
        #     high=higha
        #     while(low<=high):
        #         mid=int(low+((high-low)//2))
        #         j=int(mid%m)
        #         print(matrix[i][j])
        #         if(matrix[i][j]==target):
        #             return True
        #         else:
        #             if matrix[i][j] < target:
        #                 low=mid+1
        #             else:
        #                 high=mid-1
        #     lowa=lowa+m
        #     higha=higha+m
        # return False
        i=0
        j=m-1
        while(i<n and  j >=0):
            if matrix[i][j]==target:
                return True
            elif matrix[i][j] < target:
                i=i+1
            else:
                j=j-1
        return False

File: Search_a_2D_matrix_2.cpp
/*
Time Complexity: O(m+n)
Space Complexity: O(1)
Did this code successfully run on Leetcode :yes
Any problem you faced while coding this : no
*/

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int mid, ri = 0, ci=n-1;
        while(ci>=0 && ri<m){
            mid = matrix[ri][ci];
            if(mid==target){
                return true;
            } else if(mid>target){
                ci--;
            } else {
                ri++;
            } 
        }
        return false;
    }
};
